packages/data-context/src/sources/ProjectDataSource.ts

```diff
@@ -1,10 +1,75 @@
-import type { CodeGenType, SpecType } from '@packages/graphql/src/gen/nxs.gen'
-import { FrontendFramework, FRONTEND_FRAMEWORKS, ResolvedFromConfig, RESOLVED_FROM, SpecFileWithExtension, STORYBOOK_GLOB } from '@packages/types'
+import type { CodeGenType } from '@packages/graphql/src/gen/nxs.gen'
+import { FoundSpec, FrontendFramework, FRONTEND_FRAMEWORKS, ResolvedFromConfig, RESOLVED_FROM, SpecFileWithExtension, STORYBOOK_GLOB } from '@packages/types'
 import { scanFSForAvailableDependency } from 'create-cypress-tests'
 import path from 'path'
+import Debug from 'debug'
+import commonPathPrefix from 'common-path-prefix'
+
+const debug = Debug('cypress:data-context')
 
 import type { DataContext } from '..'
-import type { Maybe } from '../data/coreDataShape'
+
+type SpecWithRelativeRoot = FoundSpec & { relativeToCommonRoot: string }
+
+interface MatchedSpecs {
+  projectRoot: string
+  testingType: Cypress.TestingType
+  specAbsolutePaths: string[]
+  specPattern: string | string[]
+}
+export function matchedSpecs ({
+  projectRoot,
+  testingType,
+  specAbsolutePaths,
+}: MatchedSpecs): SpecWithRelativeRoot[] {
+  debug('found specs %o', specAbsolutePaths)
+
+  let commonRoot: string = ''
+
+  if (specAbsolutePaths.length === 1) {
+    commonRoot = path.dirname(specAbsolutePaths[0]!)
+  } else {
+    commonRoot = commonPathPrefix(specAbsolutePaths)
+  }
+
+  const specs = specAbsolutePaths.map((absolute) => {
+    return transformSpec(projectRoot, absolute, testingType, commonRoot)
+  })
+
+  return specs
+}
+
+export function transformSpec (projectRoot: string, absolute: string, testingType: Cypress.TestingType, commonRoot: string): SpecWithRelativeRoot {
```

Comment:
We have a similar function in `FileDataSource` called `normalizeFileToSpec`.  In the future, I think we should have one place where we do this normalizing so something to keep in mind

Reply:
Yes definitely, will look at cleaning this up soon